indexer/indexer.go

```diff
@@ -0,0 +1,383 @@
+		log.Errorw("indexer: failed to query cursor", "err", err)
+		return nil, err
+	}
+	if cfg.StartBlockNumber > cursor {
+		cursor = cfg.StartBlockNumber - 1
+	}
+
+	stakeHub, err := stakehub.New(client)
+	if err != nil {
+		log.Errorw("indexer: failed to new stakehub", "err", err)
+		return nil, err
+	}
+
+	validatorInfos, err := store.QueryValidatorInfos(ctx)
+	if err != nil {
+		log.Errorw("indexer: failed to query validator infos", "err", err)
+		return nil, err
+	}
+
+	stakeCredits := make(map[common.Address]*stakecredit.ContractWithInfo)
+	for _, v := range validatorInfos {
+		stakeCredits[v.Credit], err = stakecredit.New(v.Operator, v.Credit, client)
+		if err != nil {
+			log.Errorw("indexer: failed to new stakecredit", "err", err)
+			return nil, err
+		}
+	}
+
+	return &indexer{
+		cursor:                    cursor,
+		NumberOfBlocksForFinality: cfg.NumberOfBlocksForFinality,
+		delegators:                cfg.Delegators,
+		store:                     store,
+		client:                    client,
+		stakeHub:                  stakeHub,
+		stakeCredits:              stakeCredits,
+	}, nil
+}
+
+type indexer struct {
+	cursor                    int64
+	NumberOfBlocksForFinality int64
+	delegators                []common.Address
+	parentHeader              *types.Header
+
+	store        store.Store
+	client       *ethclient.Client
+	stakeHub     *stakehub.Contract
+	stakeCredits map[common.Address]*stakecredit.ContractWithInfo // credit address -> stake credit
+}
+
+func (i *indexer) Start(ctx context.Context) {
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := i.index(ctx)
+			if err != nil {
+				log.Errorw("indexer: failed to indexing", "err", err)
+			}
+		}
+	}
+}
+
+func (i *indexer) index(ctx context.Context) error {
+	var finalizedHeader *types.Header
+
+	if err := RtyDo(func() error {
+		var er error
+		finalizedHeader, er = i.client.FinalizedHeader(ctx, i.NumberOfBlocksForFinality)
+		return er
+	}); err != nil {
+		log.Errorw("indexer: failed to get finalized header", "err", err)
+		return err
+	}
+
+	for number := i.cursor + 1; number <= finalizedHeader.Number.Int64(); number++ {
+		var header *types.Header
+		if number == finalizedHeader.Number.Int64() {
+			header = finalizedHeader
+		} else {
+			if err := RtyDo(func() error {
+				var er error
+				header, er = i.client.HeaderByNumber(ctx, big.NewInt(number))
+				return er
+			}); err != nil {
+				log.Errorw("indexer: failed to get header by number", "number", number, "err", err)
+				return err
+			}
+		}
+
+		if i.parentHeader == nil {
+			if err := RtyDo(func() error {
+				var er error
+				i.parentHeader, er = i.client.HeaderByNumber(ctx, big.NewInt(number-1))
+				return er
+			}); err != nil {
+				log.Errorw("indexer: failed to get header by number", "number", number-1, "err", err)
+				return err
+			}
+		}
+
+		err := i.handleStakeHubLogs(ctx, header)
```

Comment:
I think it is too slow to do many filterlog query, I get two ideas to improve:
1. Query the bloom filter of the block header, if no interesting events in the header, they just skip and move to next header;
2. Utilize the range fiter logs.

I think #1 is easy.

Reply:
bloomFilter added